### PR TITLE
some state tests + fixes for closure conditions

### PR DESCRIPTION
**Pull request by yomimono**
PR is based on #174 rather than master, so the changeset is smaller than it initially appears.

Resolve a couple of TODOs:
* check ACK numbers when processing RSTs received after sending an initial SYN
* make `Tcp.State.t` abstract

In addition, a few more changes:
* replace the `Reset` state with `Closed`
* fix a bug where timers for `Fin_wait_2` would never resolve if multiple ACKs to a sent FIN were received
* process RSTs only when the sequence number exactly matches the left edge of the receive window (see [RFC 5961](https://tools.ietf.org/html/rfc5961))

And some added tests:
* a bunch of tests for `tcp/state.ml`.  Note that logic around checking seq/ack numbers is scattered between different modules in `tcp/`, so these tests can't be complete by targeting only this module.
* `Tcp.create_connection` reports failure immediately and correctly when the initial SYN receives an RST/ACK in response

### CODE REVIEW

**Comment by yomimono** [comment]
If someone could kick Travis, I'd appreciate it; 4.01 failed on a 404 downloading dependencies.

**Comment by talex5** [comment]
Kicked!

**Comment by yomimono** [comment]
Any thoughts on this?

File: lib_test/test_connect.ml [hunk 1]
```diff
@@ -83,11 +83,40 @@ module Test_connect (B : Vnetif_backends.Backend) = struct
...
+    let timeout = 15.0 in
+    Lwt.pick [
+      (Lwt_unix.sleep timeout >>= fun () ->
+       fail "connect test timed out after %f seconds" timeout) ;
+
+      (V.create_stack c backend server_ip netmask [gw] >>= fun s1 ->
+       (* register no listeners *)
+       V.Stackv4.listen s1);
+
+      (Lwt_unix.sleep 0.1 >>= fun () ->
```
Comment (talex5) [inline lib_test/test_connect.ml]: I dislike unit tests with real sleeps in them. Apart from being unreliable, when building lots of packages they all add up...
But still better than not having a test, so not a blocker for the PR.

File: tcp/state.mli [hunk 2]
```diff
@@ -40,30 +41,33 @@ type tcpstate =
...
 val state : t -> tcpstate
-val t : on_close:close_cb -> t
+val start : on_close:close_cb -> t
 
 val pp: Format.formatter -> t -> unit
 
 module Make(Time : V1_LWT.TIME) : sig
   val fin_wait_2_time : float
+  (* when in state fin_wait_2, use this as a timeout parameter  *)
   val time_wait_time : float
```
Comment (talex5) [inline tcp/state.mli]: Slightly ambiguous which value the comment refers to (`fin_wait_2_time` or `time_wait_time`).

File: tcp/state.ml [hunk 3]
```diff
@@ -123,42 +119,45 @@ module Make(Time:V1_LWT.TIME) = struct
...
-          Lwt.async (fun () -> finwait2timer t count fin_wait_2_time);
-          Fin_wait_2 count
-        else
+        if diffone b a then begin
+          Lwt.async (fun () -> finwait2timer t fin_wait_2_time);
+          Fin_wait_2
+            (* so the first time we enter fin_wait_2, we set the count to 0 and
+               spawn a waiting thread; the waiting thread only closes the
+               connection if the state still matches the number.  So if we keep
+               receiving ACKs, won't we never close? That seems incorrect. *)
```
Comment (talex5) [inline tcp/state.ml]: Out-of-date comment? Looks like you removed the counter.
@balrajsingh any idea what this counter scheme was for?

**Comment by talex5** [comment]
I'm not too familiar with the TCP state model, but the parts of the patch I understand look good :-) Would like to understand what the fin counter stuff was for though - presumably it was added for a reason, even if it's wrong in other ways.

**Comment by yomimono** [comment]
I'll separate out a few things and submit a new PR which doesn't conflict with the just-merged #182 .

**Pull request closed without merging**